feat: Search field under all tabs in Variable Table (Issue #287)

not-Ian opened the pull request:
1. I renamed a bunch of variables and display names. (Tabs on the 'Variables' dialog had slightly incorrect wording - IE: "Lines angles" was renamed to 'Line angles". Also renamed corresponding variables/function names)

2. I created a private function on 'dialogIncrements.ccp/.h' that setup all the search/findPrevious/findNext functionality for every tab. In short:
   - It creats a array of tuples, of size 7. Each tuple corresponding to a tab on the Tables of Variables dialog, and each containing a: (VTableSearch, QLineEdit, QToolButton, QToolButton)
   - It iterates through the array, and hooks up the VTableSearch, QLineEdit, and QToolButtons accordingly
3. We now refresh all 7 VTableSearch objects and their corresponding tableViews when increments are updated/changed (in functions 'UpdateTree' starting on line 546, and 'FullUpdateFromFile' starting on line 608)

One thing remaining, I'm not sure if I finished the UI file properly?? I had to 'Break Layout' to add the new widgets... I'm not sure how to 'unbreak' them??

not-Ian committed:
1. I renamed a bunch of variables and display names. (Tabs on the 'Variables' dialog had slightly incorrect wording - IE: "Lines angles" was renamed to 'Line angles". Also renamed corresponding variables/function names)
2. I created a private function on 'dialogIncrements.ccp/.h' that setup all the search/findPrevious/findNext functionality for every tab. In short:
   - It creats a array of tuples, of size 7. Each tuple corresponding to a tab on the Tables of Variables dialog, and each containing a: (VTableSearch, QLineEdit, QToolButton, QToolButton)
   - It iterates through the array, and hooks up the VTableSearch, QLineEdit, and QToolButtons accordingly
3. We now refresh all 7 VTableSearch objects when a new file is loaded (in functions 'UpdateTree' starting on line 546, and 'FullUpdateFromFile' starting on line 608) (Do we really need to do this? Can this window be open when a new file loads in?)

One thing remaining, I'm not sure if I finished the UI file properly?? I had to 'Break Layout' to add the new widgets... I'm not sure how to 'unbreak' them??

DSCaskey commented:
>

   - 1. I renamed a bunch of variables and display names. (Tabs on the
   'Variables' dialog had slightly incorrect wording - IE: "Lines angles" was
   renamed to 'Line angles". Also renamed corresponding variables/function
   names)

Glad to see you watched my video and fixed the Tab labels.;)

On Sat, Sep 19, 2020, 2:19 PM Ian Cunningham <notifications@github.com>
wrote:

> …riables' dialog had slightly incorrect wording - IE: "Lines angles" was
> renamed to 'Line angles". Also renamed corresponding variables/function
> names)
>
>    1. I created a private function on 'dialogIncrements.ccp/.h' that
>    setup all the search/findPrevious/findNext functionality for every tab. In
>    short:
>       - It creats a array of tuples, of size 7. Each tuple corresponding
>       to a tab on the Tables of Variables dialog, and each containing a:
>       (VTableSearch, QLineEdit, QToolButton, QToolButton)
>       - It iterates through the array, and hooks up the VTableSearch,
>       QLineEdit, and QToolButtons accordingly
>    2. We now refresh all 7 VTableSearch objects when a new file is loaded
>    (in functions 'UpdateTree' starting on line 546, and 'FullUpdateFromFile'
>    starting on line 608) (Do we really need to do this? Can this window be
>    open when a new file loads in?)
>
> One thing remaining, I'm not sure if I finished the UI file properly?? I
> had to 'Break Layout' to add the new widgets... I'm not sure how to
> 'unbreak' them??
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/FashionFreedom/Seamly2D/pull/312
> Commit Summary
>
>    - 1. I renamed a bunch of variables and display names. (Tabs on the
>    'Variables' dialog had slightly incorrect wording - IE: "Lines angles" was
>    renamed to 'Line angles". Also renamed corresponding variables/function
>    names)
>
> File Changes
>
>    - *M* src/app/seamly2d/dialogs/dialogincrements.cpp
>    <https://github.com/FashionFreedom/Seamly2D/pull/312/files#diff-a3b92ef0c9da96b7eacc96752e78b16e>
>    (297)
>    - *M* src/app/seamly2d/dialogs/dialogincrements.h
>    <https://github.com/FashionFreedom/Seamly2D/pull/312/files#diff-0795bdca566caa96740b5f942a958c8b>
>    (35)
>    - *M* src/app/seamly2d/dialogs/dialogincrements.ui
>    <https://github.com/FashionFreedom/Seamly2D/pull/312/files#diff-38460864dc8f80a3fe44a36e686b9e2b>
>    (1620)
>
> Patch Links:
>
>    - https://github.com/FashionFreedom/Seamly2D/pull/312.patch
>    - https://github.com/FashionFreedom/Seamly2D/pull/312.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/312>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHTXADV5G2YAL6SQDZFC66DSGTY4XANCNFSM4RTE7ZCQ>
> .
>

slspencer commented:
Hey @not-Ian This PR was reverted. There's a bug filed against it, issue #313.  Did this run on your local PC without crashing? How have you defined your compiler kit in Qt?

not-Ian commented:
Yes, I ran it on my machine. Maybe I made further changes without testing?
I know I saw the search function working.

I'll look into it and figure out what's going on with the UI.

Sorry about this

On Mon, Sep 21, 2020, 10:34 AM Susan Spencer <notifications@github.com>
wrote:

> Hey Ian, this PR was reverted. There's a bug filed against it, issue #313
> <https://github.com/FashionFreedom/Seamly2D/issues/313>. Did this run on
> your local PC without crashing? How have you defined your compiler kit in
> Qt?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-696173235>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEJNGCU2GWUWSFN3ODSV23SG5TIDANCNFSM4RTE7ZCQ>
> .
>

DSCaskey commented:
Hey Ian... what Qt ver did you build with, and what OS did you build and
run on?

On Mon, Sep 21, 2020, 11:53 AM Ian Cunningham <notifications@github.com>
wrote:

> Yes, I ran it on my machine. Maybe I made further changes without testing?
> I know I saw the search function working.
>
> I'll look into it and figure out what's going on with the UI.
>
> Sorry about this
>
> On Mon, Sep 21, 2020, 10:34 AM Susan Spencer <notifications@github.com>
> wrote:
>
> > Hey Ian, this PR was reverted. There's a bug filed against it, issue #313
> > <https://github.com/FashionFreedom/Seamly2D/issues/313>. Did this run on
> > your local PC without crashing? How have you defined your compiler kit in
> > Qt?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-696173235
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACEJNGCU2GWUWSFN3ODSV23SG5TIDANCNFSM4RTE7ZCQ
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-696206484>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHTXADU6XPNDCHDU74PBLALSG5ZG7ANCNFSM4RTE7ZCQ>
> .
>

DSCaskey commented:
I looked at the Dialog Increments code and I'm sort of baffled why there is
a need to use tupels & have 7 lineEdits, and findPrevious / findNext
toolButtons?  Seems over complicated. I would just use one search bar (move
it out of  the QTabWidget) and switch the VTableSearch depending on which
tab is selected.

[image: search.png]

 I noticed the search bar keeps moving around depending on which tab is
selected... kind of wonky. Another side effect from not being in a layout.

Also... is there a reason why the event filter is only installed on the
lineEditFindIncrements and not the other 6 lineEdits you added?

    ui->lineEditFindIncrements->setClearButtonEnabled(true);
    ui->lineEditFindIncrements->installEventFilter(this);

That said the problem is here... due to using unsigned int
for(unsigned int tableSearchItemIndex = 0; tableSearchItemIndex <
sizeof(tableSearchItems); tableSearchItemIndex = tableSearchItemIndex + 1 )

Try:
for(int tableSearchItemIndex = 0; tableSearchItemIndex < 7 ;
++tableSearchItemIndex)

On Mon, Sep 21, 2020 at 2:12 PM Douglas Caskey <dscaskey@gmail.com> wrote:

> Hey Ian... what Qt ver did you build with, and what OS did you build and
> run on?
>
> On Mon, Sep 21, 2020, 11:53 AM Ian Cunningham <notifications@github.com>
> wrote:
>
>> Yes, I ran it on my machine. Maybe I made further changes without testing?
>> I know I saw the search function working.
>>
>> I'll look into it and figure out what's going on with the UI.
>>
>> Sorry about this
>>
>> On Mon, Sep 21, 2020, 10:34 AM Susan Spencer <notifications@github.com>
>> wrote:
>>
>> > Hey Ian, this PR was reverted. There's a bug filed against it, issue
>> #313
>> > <https://github.com/FashionFreedom/Seamly2D/issues/313>. Did this run
>> on
>> > your local PC without crashing? How have you defined your compiler kit
>> in
>> > Qt?
>> >
>> > —
>> > You are receiving this because you authored the thread.
>> > Reply to this email directly, view it on GitHub
>> > <
>> https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-696173235
>> >,
>> > or unsubscribe
>> > <
>> https://github.com/notifications/unsubscribe-auth/ACEJNGCU2GWUWSFN3ODSV23SG5TIDANCNFSM4RTE7ZCQ
>> >
>> > .
>> >
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-696206484>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AHTXADU6XPNDCHDU74PBLALSG5ZG7ANCNFSM4RTE7ZCQ>
>> .
>>
>

not-Ian commented:
So, I haven't once actually gotten a crash. I wonder if you had anything special going on with your pattern that's making that crash? This is my 2nd time programming in C++, so I might be missing some nuance in how it does things. 

Anyways, yes, you're right, the 7 line edits is overkill. I think I've changed the UI properly now and simplified it? I pushed some changes. There might still be a few small broken pieces in the UI? Items appear with a red border, but nothing appears to be "Broken" in the widget tree on the right. It appears to be resizing properly.

Also, searching is now a little funky, I couldn't quite figure out what was going wrong. Something about the VTableSearch calling "clear" when "findNext" or "findPrevious" is called, disabling both findNext/findPrevious buttons. If I could get a second pair of eyes on this it would be great. Again, 2nd time ever doing C++ (first time was with my first SeamlyMe PR :P) 

Thanks

not-Ian commented:
Oh, btw, building onto Windows with QTCreator 4.13.0

DSCaskey commented:
Hey Ian...

It was nothing to do with the pattern I was using... it was crashing even
with an empty new pattern, Anyhow it's a moot point as you removed the for
loop that was causing the problem.

> There might still be a few small broken pieces in the UI? Items appear
with a red border

Normal. The red outline indicates the widgets are in a layout. If you
follow the Object Inspector window you can see how the widgets are  nested
and in what layout etc. If you wanted you could combine the 2 horizontal
layouts by moving the 2 arrow buttons into the other layout... just drag
them over, and then delete the other layout & expander. Not the end of
the world as it is though. That said, the Dialog is now not cutting off
when resized. :)

[image: details.png]

>Also, searching is now a little funky,

No doubt. On the right track... but for one you're adding more connections
by calling SetupTableSearch()  every time the tab is switched.
SetupTableSearch()  should only be handled once - as if it was in the
constructor itself. Need to have a separate method that switches the search
pointer. and connects there instead of SetupTableSearch(). If you search
the code you will find that TabChanged(int index) is used elsewhere...
would probably be good to stick with that method name for consistency.
Which brings up the issue... currentTable needs to be a member of
the DialogIncrements to be in scope outside of  SetupTableSearch().

Also you're refreshing the lineEditFind from all the VTableSearch's and not
just the currentTable in FullUpdateFromFile() and UpdateTree().
    searchIncrements->RefreshList(ui->lineEditFind->text());
    ....
    searchArcRadiuses->RefreshList(ui->lineEditFind->text());

 I assume you just need to refresh the current table?

   currentTable->RefreshList(ui->lineEditFind->text());

Hope this helps.



On Wed, Sep 23, 2020, 1:07 AM Ian Cunningham <notifications@github.com>
wrote:

> Oh, btw, building onto Windows with QTCreator 4.13.0
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/FashionFreedom/Seamly2D/pull/312#issuecomment-697136819>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHTXADXKUS2XU7FHYQ2O6UDSHF7CHANCNFSM4RTE7ZCQ>
> .
>

not-Ian commented:
I finally got it working. This PR has been weighing on my conscious like a homunculus at night. The missing code a reflection of my faults, and the errors in this world. FIND NEXT. FIND PREVIOUS. These words echo in my head. I can't find anything lately. This search bar's cursor blinks like the seconds of the day. Parables worldwide throughout history written to find something within - when will a Seamly2D user be able to find their measurements? When will an arc of sunlight shine down on this arc radius?? Could the dialogue increments UI EVER BE SO GRACIOUS AS TO REVEAL ITS LABRYTHIN OF VARIABLES???

Anyways, I definitely had to take the currentTable and make it class-wide scope. Also I dropped the disabling/enabling of findNext/findPrevious if nothing is found. Wasn't working as expected, seems like a nice-to-have.
